### DISCUSSION TITLE
Tox 3 vs Tox 4: Tox 4 doesn't need pyproject.toml anymore? What changed?

### DISCUSSION THREAD

**Discussion by samuelwhiskeyjohnson** [tox-dev/tox#3296] [Q&A] [answered]
I was reading the book Python Testing with Pytest, which I think it uses tox 3.

How come running in `tox -e py310 -c tox_bad.ini` with tox 3.24.5 causes error but running with tox 4.15.1 doesn't cause the error?

**hello.py**
```
print("Hello, World!")
```

**test_hello.py**
```
from subprocess import run


def test_hello():
    result = run(["python", "hello.py"], capture_output=True, text=True)
    output = result.stdout
    assert output == "Hello, World!\n"

```

**tox_bad.ini**
```
[tox]
envlist = py39, py310
[testenv]
deps = pytest
commands = pytest
[pytest]
```

Running `tox -e py310 -c tox_bad.ini` causes this error running with tox 3.24.5 
**The error**
```
ERROR: No pyproject.toml or setup.py file found. The expected locations are:
/path/to/code/ch12/script/pyproject.toml or /path/to/code/ch12/script/setup.py
You can
1. Create one:
https://tox.readthedocs.io/en/latest/example/package.html
2. Configure tox to avoid running sdist:
https://tox.readthedocs.io/en/latest/example/general.html
3. Configure tox to use an isolated_build
```

Running `tox -e py310 -c tox_bad.ini` with tox==4.15.1 runs fine.

**Comment by gaborbernat** [comment, marked as answer]
Because we default to package skip if file is missing https://github.com/tox-dev/tox/blob/main/src/tox/tox_env/python/virtual_env/runner.py#L36

**Reply by samuelwhiskeyjohnson** [reply]
If the files (pyproject.toml, setup.py, setup.cfg)  are missing, is that the same as setting `skipsdist = true`?
